src/java.rmi/share/classes/sun/rmi/transport/tcp/TCPDirectSocketFactory.java

```diff
@@ -25,19 +25,38 @@
 package sun.rmi.transport.tcp;
 
 import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
 import java.net.Socket;
+import java.net.SocketAddress;
 import java.net.ServerSocket;
 import java.rmi.server.RMISocketFactory;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
 
 /**
  * RMIDirectSocketFactory creates a direct socket connection to the
  * specified port on the specified host.
  */
 public class TCPDirectSocketFactory extends RMISocketFactory {
 
+    @SuppressWarnings("removal")
+    private static final int connectTimeout =    // default 1 minute
+        AccessController.doPrivileged((PrivilegedAction<Integer>) () ->
+            Integer.getInteger("sun.rmi.transport.tcp.initialConnectTimeout", 60 * 1000).intValue());
```

Comment:
An exception here will prevent the class from being initialized, any subsequent attempts to use the class will then produce hard to diagnose error. If the class is not loaded by the main thread, such exception/error could be swallowed unless an uncaught exception handler was registered. I would suggest implementing the logic in a static block, catch any exception and revert to default behaviour (i.e. 0) if the value supplied for the property is not an integer, or not a valid integer, etc...

Reply:
IIRC, the default agent uses / may use its own socket factories - are we still coming here even with those factories?

Reply:
> An exception here will prevent the class from being initialized...

Maybe we can break it, but this new property is following the same pattern I think as the neighbouring file src/java.rmi/share/classes/sun/rmi/transport/tcp/TCPChannel.java when it reads the system props.  I see Integer.getInteger handles the obvious Exceptions, so specifying a strange value does not immediately break us.

If there's more protection needed then we have various other places to apply it that might need separate follow-up if you think there's a case?

Reply:
> IIRC, the default agent uses / may use its own socket factories - are we still coming here even with those factories?

We do get here, yes (not saying you can't customise your way out of getting here).  This is a stack from a test I was experimenting with, when it did see the timeout:

```
---------System.out:(4/132)----------
JMX URL = service:jmx:rmi://x
expectTimeout = true
sun.rmi.transport.tcp.initialConnectTimeout = 1
Test passed
----------System.err:(24/1791)----------
java.rmi.ConnectIOException: Exception creating connection to: x.x.x.x; nested exception is:
        java.net.SocketTimeoutException
        at java.rmi/sun.rmi.transport.tcp.TCPEndpoint.newSocket(TCPEndpoint.java:637)
        at java.rmi/sun.rmi.transport.tcp.TCPChannel.createConnection(TCPChannel.java:217)
        at java.rmi/sun.rmi.transport.tcp.TCPChannel.newConnection(TCPChannel.java:204)
        at java.rmi/sun.rmi.server.UnicastRef.invoke(UnicastRef.java:134)
        at java.management.rmi/javax.management.remote.rmi.RMIServerImpl_Stub.newClient(RMIServerImpl_Stub.java:85)
        at java.management.rmi/javax.management.remote.rmi.RMIConnector.getConnection(RMIConnector.java:2106)
        at java.management.rmi/javax.management.remote.rmi.RMIConnector.connect(RMIConnector.java:321)
        at java.management/javax.management.remote.JMXConnectorFactory.connect(JMXConnectorFactory.java:270)
        at java.management/javax.management.remote.JMXConnectorFactory.connect(JMXConnectorFactory.java:229)
        at RMIConnectionTimeoutTest.main(RMIConnectionTimeoutTest.java:65)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
        at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: java.net.SocketTimeoutException
        at java.base/java.net.SocksSocketImpl.remainingMillis(SocksSocketImpl.java:112)
        at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
        at java.base/java.net.Socket.connect(Socket.java:752)
        at java.rmi/sun.rmi.transport.tcp.TCPDirectSocketFactory.createSocket(TCPDirectSocketFactory.java:57)
        at java.rmi/sun.rmi.transport.tcp.TCPEndpoint.newSocket(TCPEndpoint.java:619)
        ... 13 more
STATUS:Passed.
```

Reply:
>  I see Integer.getInteger handles the obvious Exceptions, so specifying a strange value does not immediately break us.

Oh - right. It's `Integer.getInteger`, not `Integer.parseInt` Ok, then - I withdraw my first comment :-)

Reply:
> This is a stack from a test I was experimenting with, when it did see the timeout:

Ah - that's for testing with a particular test. Si the question now is:

Should the RMISocketFactories provided / implemented in the JMX implementation - such as those used by the JMX default agent, also honour this system property?

Reply:
(Look for socket factories in the module `jdk.management.agent`)

Reply:
OK yes, we also have: java.rmi/share/classes/javax/rmi/ssl/SslRMIClientSocketFactory.java
with its own createSocket(String host, int port) method.  This is used if we use JMX over SSL.

So SslRMIClientSocketFactory could specifically implement the connect timeout.

Next q, should it?  8-)

The reported hang and those I have seen in testing have only been in:
sun.rmi.transport.tcp.TCPDirectSocketFactory.createSocket calling Socket.init.

javax/rmi/ssl/SslRMIClientSocketFactory.java reads some properties named "javax.rmi.ssl.client...."
so it would be odd for it to read "sun.rmi.transport.tcp.initialConnectTimeout" I was proposing here.

It could implement "javax.rmi.ssl.client.initialConnectTimeout", or we could leave SSL alone for now, possibly handling it in a separate issue if it's wanted.

Reply:
OK - sounds good. Meanwhile I had a look at the custom RMI Socket Factories used by the JMX Agent, and these are actually RMIServerSocketFactories, so having a timeout for connect there probably makes no sense.

Reply:
Thanks, yes so JMX SSL clients are without socket connect timeout, good to have that documented.  I logged a parallel RFE for SslRMIClientSocketFactory, which should wait for now as it seems wrong to go changing that right now without knowing if it's ever been an issue for anybody:
[JDK-8320703](https://bugs.openjdk.org/browse/JDK-8320703): JMX SSL RMI Socket connection timeout cannot be changed

Reply:
wrt to the property name initialConnectTimeout might infer that it is an initial value which can be subsequentually modified, but that is not possible. As such, would sun.rmi.transport.tcp.connectTimeout be more appropriate -- dropping the initial ?

If the connectTimeout initialization code was placed in a static method, it could directly unit testable :-) (if such a test was desirable for coverage completeness)

Reply:
On the Property name: there is an existing System Property "sun.rmi.transport.connectionTimeout" which is a 15-second idle timeout.  Having a "connectionTimeout" and this new one as  "connectTimeout" could be confusing, even in different but very similar package names, hence naming this "initialConnectTimeout".
(So "initial" to signal that it's the initial connect of a socket, different to the existing "connectionTimeout".) 
It does not seem like the kind of system property a user would expect to be read again on every usage, I think, let me know if you see that as a problem.

I am hoping that as we already have various properties fetched in the same manner, we don't see the need to pursue a test that the value is fetched and passed on.

Reply:
If you named it "tcpConnectTimeout" then you could name the (future) property for [JDK-8320703](https://bugs.openjdk.org/browse/JDK-8320703: JMX SSL RMI Socket connection timeout cannot be changed) "tlsConnectTimeout"...

Reply:
Yes, that could work 8-)